### PR TITLE
[engraving] Added iwriter interface and msc saver

### PR DESCRIPTION
**Pull request by igorkorsukov**

### CODE REVIEW

**Review by RomanPudashkin — APPROVED** [review]

File: src/engraving/rw/rwregister.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    } else if (version < 400 || MScore::testMode) {
+        return std::make_shared<compat::Read302>();
+    }
+
+    return std::make_shared<rw400::Read400>();
+}
+
+IReaderPtr RWRegister::latestReader()
+{
+    return reader(-1);
```
Comment (cbjeukendrup) [inline src/engraving/rw/rwregister.cpp:49]: Wouldn't this return the 206 reader? (since `-1 <= 207`)

**Comment by cbjeukendrup** [comment]
Why is the writer in a versioned namespace? We always only support writing for the latest file format. Or are there plans to change this? I.e. add a new feature to save scores in a way that older versions can open them. I don't think such a feature would be worth the effort...

File: src/engraving/rw/rwregister.cpp (hunk 1 above)
Comment (igorkorsukov) [reply src/engraving/rw/rwregister.cpp:49]: hmm, I fixed that, but it looks like I rolled back the last edit. Thank you

**Comment by igorkorsukov** [comment]
> Why is the writer in a versioned namespace? We always only support writing for the latest file format. Or are there plans to change this? I.e. add a new feature to save scores in a way that older versions can open them. I don't think such a feature would be worth the effort...

Good question.
The main idea was that the writer was next to the reader, so that it was clear that they should correspond to each other and it was easier to check that they correspond to each other (when it is in the same folder).

I also don’t see the point in supporting several formats for writing, the efforts exceed the profit (only if for a test in the transition period)

I agree that a writer can do it in a separate folder in a namespace without a version (or namespace latest).
How would you do it and why?

**Comment by igorkorsukov** [comment]
Let me put the writer in a separate folder and namespace in the next PR

**Comment by cbjeukendrup** [comment]
Maybe we could have a "latest" namespace, and put the latest reader and the writer in that folder. 
(I proposed that "latest" namespace earlier, but for a different reason; it turned out that the original reason did not really have advantages, but no real disadvantages either, so maybe it's still advantageous for a different reason, namely for having the reader and writer together.

**Comment by igorkorsukov** [comment]
I think it's better not to enter `latest`, it will complicate the addition of a new version of the format. And this will contribute to the fact that they will not add a new version, but will change `latest` (as now), as a result we will get complexity and a mess.

**Pull request merged**